{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of sequences in the file: 6\n",
      "\n",
      "SequenceID     #Exons         #Introns       AvgExonLength  AvgIntronLeng  %A in Exon     %C In Exon     %G In Exon     %T In Exon     %A in Intron   %C In Intron   %G In Intron   %T In Intron   \n",
      ">HCN1          8              7              1241.62        61642.86       30.25          20.22          20.0           29.53          32.1           16.94          17.69          33.27          \n",
      ">HCN2          8              7              427.5          3408.43        15.12          38.71          32.11          14.06          17.53          28.71          33.35          20.42          \n",
      ">KCNA1         2              1              1309.0         4062.0         27.73          24.41          21.28          26.59          28.39          20.83          20.7           30.08          \n",
      ">KCNA2         5              4              391.8          8984.0         24.4           26.9           24.2           24.5           25.49          23.22          23.41          27.88          \n",
      ">KCNS3         3              2              780.33         25975.5        23.88          23.96          25.54          26.61          25.37          19.25          21.91          33.46          \n",
      ">KIR3          3              2              209.33         1668.0         20.54          30.1           25.0           24.36          32.22          20.23          27.61          19.93          \n",
      "\n",
      "DNAstats.txt file is created.\n"
     ]
    }
   ],
   "source": [
    "# Open the fasta file in read mode\n",
    "file = open(\"FinalProject.fasta\", 'r')\n",
    "\n",
    "sequences = {}\n",
    "\n",
    "# Parse the fasta file and extract sequences\n",
    "for line in file:\n",
    "    line = line.strip()\n",
    "    if line.startswith('>'):\n",
    "        seq_id = line\n",
    "        sequences[seq_id] = {}\n",
    "        sequences[seq_id]['seq'] = ''\n",
    "    else:\n",
    "        sequences[seq_id]['seq'] += line\n",
    "file.close()\n",
    "\n",
    "# Display number of sequences\n",
    "print(\"Number of sequences in the file:\", len(sequences), end=\"\\n\\n\")\n",
    "\n",
    "\n",
    "for seq_id, seq in sequences.items():\n",
    "    exons = []\n",
    "    introns = [] \n",
    "    exon, intron = '', ''\n",
    "   \n",
    "    for i, k in enumerate(seq['seq']):\n",
    "        if k.isupper(): \n",
    "            if intron: \n",
    "                introns.append(intron)\n",
    "                intron = ''\n",
    "            exon += k\n",
    "        elif k.lower() and (exons or exon): \n",
    "            if exon: \n",
    "                exons.append(exon)\n",
    "                exon = ''\n",
    "            intron += k\n",
    "    seq['exons'] = exons\n",
    "    seq['introns'] = introns\n",
    "    seq['AvgExonLength'] = round(sum(len(exon) for exon in exons)/len(exons),2)\n",
    "    seq['AvgIntronLeng'] = round(sum(len(intron) for intron in introns)/len(introns),2)\n",
    "    seq['A_in_Exon'] = round((''.join(exons).count('A')/len(''.join(exons)))*100, 2)\n",
    "    seq['C_in_Exon'] = round((''.join(exons).count('C')/len(''.join(exons)))*100, 2)\n",
    "    seq['G_in_Exon'] = round((''.join(exons).count('G')/len(''.join(exons)))*100, 2)\n",
    "    seq['T_in_Exon'] = round((''.join(exons).count('T')/len(''.join(exons)))*100, 2)\n",
    "    seq['A_in_Intron'] = round((''.join(introns).count('a')/len(''.join(introns)))*100, 2)\n",
    "    seq['C_in_Intron'] = round((''.join(introns).count('c')/len(''.join(introns)))*100, 2)\n",
    "    seq['G_in_Intron'] = round((''.join(introns).count('g')/len(''.join(introns)))*100, 2)\n",
    "    seq['T_in_Intron'] = round((''.join(introns).count('t')/len(''.join(introns)))*100, 2)\n",
    "\n",
    "# Header\n",
    "headers = [\"SequenceID\", \"#Exons\", \"#Introns\", \"AvgExonLength\", \"AvgIntronLeng\", \"%A in Exon\",\n",
    "           \"%C In Exon\", \"%G In Exon\", \"%T In Exon\", \"%A in Intron\", \"%C In Intron\", \"%G In Intron\",\n",
    "           \"%T In Intron\"]\n",
    "\n",
    "# Print header\n",
    "print((\"{:<15}\"*len(headers)).format(*headers))\n",
    "# Loop through data stored in sequences dictionary\n",
    "# and print the data\n",
    "for seq_id, seq in sequences.items():\n",
    "    print((\"{:<15}\"*len(headers)).format(seq_id, len(seq['exons']), len(seq['introns']), seq['AvgExonLength'], \n",
    "        seq['AvgIntronLeng'], seq['A_in_Exon'], seq['C_in_Exon'], seq['G_in_Exon'], seq['T_in_Exon'], \n",
    "        seq['A_in_Intron'], seq['C_in_Intron'], seq['G_in_Intron'], seq['T_in_Intron']))\n",
    "\n",
    "# Open DNAstats.txt in writing mode\n",
    "output = open('DNAstats.txt', 'w')\n",
    "# Write header\n",
    "output.write('\\t'.join(headers)+\"\\n\")\n",
    "# Loop through data stored in sequences dictionary\n",
    "# and write the data to the file\n",
    "for seq_id, seq in sequences.items():\n",
    "    output.write('\\t'.join(map(str,[seq_id, len(seq['exons']), len(seq['introns']), seq['AvgExonLength'], \n",
    "        seq['AvgIntronLeng'], seq['A_in_Exon'], seq['C_in_Exon'], seq['G_in_Exon'], seq['T_in_Exon'], \n",
    "        seq['A_in_Intron'], seq['C_in_Intron'], seq['G_in_Intron'], seq['T_in_Intron']]))+\"\\n\")\n",
    "\n",
    "print(\"\\n{} file is created.\".format(output.name))\n",
    "output.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
